### ISSUE TITLE
Improve performance : compute packages and weight ahead of time (in the database ?) to reduce the numbers of queries

### ISSUE THREAD

**Issue by Atala** [coopcycle/coopcycle-web#4401]
### What is the issue?

The Task object is linked to `packages` (many to many from package to task) and has `weight`. However now we want to display the following for delivery/orders :
 - `Pickup` : sum of weight of all the dropoffs, "sum of packages" of all the dropoffs
 - `Dropoff` : weight of the task (easy ^^), nicely serialized packages (like a JSON `{type: XL, quantity: 2}`)

However for now this is not precomputed, so we compute it each time we want to serialize a task in the normalizer ([ref](https://github.com/coopcycle/coopcycle-web/blob/master/src/Serializer/TaskNormalizer.php#L78-L109))

so for a dispatch panel with 300 tasks it adds ~600 queries. I created a custom endpoint to get all the tasks, so it will prefetch and "annotate the tasks" with the packages and weight data (https://github.com/coopcycle/coopcycle-web/blob/master/src/Action/Task/Collection.php#L19-L73) and it worked (loading time from 15s to 1s).

However the issue remains in part of the code. For example when you fire the `task:assigned` event  you may serialize ~30 tasks (when assigning a tour for example) and then you get ~60 queries more.

So I do think these fields (JSON packages and weight) should be precomputed when we load the tasks objects. Now they are two solutions : 
 - doctrine listener on task collections to get all of the thing (dont know how it would work in practice ??)
 - computed field in database and/or database view and/or stored procedures -> i think this is the way to go, like having an extra field on task in the database that is computed by postgres

What do you think? what would be the good solution?

**Comment by alexsegura** [comment]
I think what impacts performance is not really the number of SQL queries, but the hydration of PHP objects. Serialization takes time, because not only it needs to load all the related objects (which happens at serialization time, because Doctrine load lazily those objects), but it also needs to hydrate them with all their internal dependencies, filling up the memory and slowing down the whole thing. 

I think we should keep the serializer simple, it should just serialize blindly what has been computed at previous stages. Maybe we could use a [custom Doctrine hydrator](https://medium.com/@thovandeth/creating-a-custom-hydrator-in-doctrine-and-symfony-for-enhanced-performance-with-dtos-f9f4fce8f035), that loads the packages and stuff as simple objects without any dependencies.

**Comment by Atala** [comment]
- I do agree it could be a way to improve further things for the collections endpoint, as we can see here it spends a lot of time in PHP doing i dont know what, so that may be hydratation [https://app.datadoghq.com/apm/trace/666fec9d00000000b1750133ed218468?colorBy=service&env=prod&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=12787128039521748072&spanViewType=metadata&timeHint=1718611101312.0288&view=spans](https://app.datadoghq.com/apm/trace/666fec9d00000000b1750133ed218468?colorBy=service&env=prod&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=12787128039521748072&spanViewType=metadata&timeHint=1718611101312.0288&view=spans)
- also in this case using pagination to get tasks in the dashboard could improve things?
- I do think the number of SQL queries as also an impact, as if you count 10-30ms per package/weight query and you do it on serializer level then it goes up quite fase when the tasks collection size increase

**Comment by vladimir-8** [comment]
> However the issue remains in part of the code. For example when you fire the task:assigned event you may serialize ~30 tasks (when assigning a tour for example) and then you get ~60 queries more.

Maybe a solution for this is getting rid of the `assigned_to` column in the `Task` table? Why do we need it if a task eventually belongs to a task list (either directly or via a tour) and a task list is assigned to a courier? In that case `task:assigned` would mean an assignment either to a task list or to a tour and if we you assign a tour we will need a `tour:assigned` event

**Comment by Atala** [comment]
yes getting rid of this field would be my/the ultimate goal 💯 (I have been thinking about it for a long time ^^). I don't think it is related to this performance issue but more tech debt and internal coherence and some bugs.

the current setup is : 
 - most of the app code relies on task.assignedTo (for ex when you dispatch from the app you modify this field)
 - most of the web endpoints relies on tasklist.items but some old code still read task.assignedTo
 - EntityChangeSetProcessor class is the huge machinery to sync them together

this causes several problems for example : 
 - task.assignedTo is not tied to a day, so tasks spawning several days are not assignable on several days w.o. breaking things
 - very complicated listener to sync all of this
 - very fragile code on top of that

we can discuss it, I guess in another issue/place. I think there is path forward but it needs time and thinking. I have some ideas :)
